src/schema-parser.js

```diff
@@ -25,13 +25,15 @@ export default class SchemaParser {
     const selfURL = extract(links, 'self.href');
     const parsedSelfURL = parseJsonApiUrl(selfURL);
     if (Object.keys(parsedSelfURL.query.fields).length) {
-      parsedSelfURL.query.fields = [];
+      const completeFieldsetUrl = Object.assign({}, parsedSelfURL, {query: Object.assign({}, parsedSelfURL.query, {fields: []})});
       return this.parse(
-        Document.parse(await request(compileJsonApiUrl(parsedSelfURL))),
+        Document.parse(await request(compileJsonApiUrl(completeFieldsetUrl))),
         forPath,
       );
     }
-    return this.inferSchema(root, forPath);
+    const baseResourceURL = compileJsonApiUrl(Object.assign({}, parsedSelfURL, {protocol: 'inferred:', query: {}}));
+    const inferredSchema = this.inferSchema(root, forPath);
+    return !forPath.length ? this.mergeWithCachedInference(baseResourceURL, inferredSchema) : inferredSchema;;
```

Comment:
This is the important part, if `inferredSchema` is null it will merge in the cached schema.